### PR TITLE
 Added hanainstance restart command

### PR DESCRIPTION
**Pull request by PakDLiu**
Added hanainstance restart command

NOTE:
The HANA RP restart hanainstance REST API is a POST with no body and expects the Content-Type to be set.
Swagger does not add the Content-Type in the header if there is no body, this is by design.
A custom header is added as a mitigation. Will need to look into the HANA RP expecting Content-Type to be set.

### CODE REVIEW

File: azext_hanaonazure/_help.py [hunk 1]
```diff
@@ -19,4 +19,9 @@
 helps['hanainstance list'] = """
     type: command
     short-summary: List SAP HANA Instances.
+"""
+
+helps['hanainstance restart'] = """
+    type: command
+    short-summary: Restart a SAP HANA Instance.
 """
```
Comment (jeffaco) [inline azext_hanaonazure/_help.py]: Should this have a newline at the end of the file?

File: HISTORY.rst [hunk 2]
```diff
@@ -3,6 +3,11 @@
 Release History
 ===============
 
+0.1.5 (2018-05-18)
```
Comment (lagalbra) [inline HISTORY.rst]: Copy-pasta: date should be today's date

File: azext_hanaonazure/_help.py (hunk 1 above)
Comment (pabowers) [inline azext_hanaonazure/_help.py]: nit: newline

File: azext_hanaonazure/azext_metadata.json [hunk 3]
```diff
@@ -1,4 +1,4 @@
 {
   "azext.minCliCoreVersion": "2.0.25",
-  "version": "0.1.1"
+  "version": "0.1.5"
 }
```
Comment (pabowers) [inline azext_hanaonazure/azext_metadata.json:4]: nit: newline

File: azext_hanaonazure/commands.py [hunk 4]
```diff
@@ -12,4 +12,5 @@
 def load_command_table(self, _):
     with self.command_group('hanainstance', client_factory=cf_hanainstance_groups) as g:
         g.custom_command('list', 'list_hanainstance')
-        g.custom_command('show', 'show_hanainstance', exception_handler=empty_on_404)
\ No newline at end of file
+        g.custom_command('show', 'show_hanainstance', exception_handler=empty_on_404)
+        g.custom_command('restart', 'restart_hanainstance')
```
Comment (pabowers) [inline azext_hanaonazure/commands.py]: nit: newline

**Review by jeffaco — APPROVED** [review]
Customary to have newlines at the end of .py files, even if they aren't mandatory.

You should combine your second commit into the first; a commit to change a comment is insignificant, and just belongs with the major change here.

It looks like this has no unit test framework? I'm a huge believer in unit tests, and would like to see a framework developed where this can be routinely tested (and hooked into CI systems like Travis, etc). Won't require that, but strongly recommend it.

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py [hunk 5]
```diff
@@ -237,3 +238,56 @@ def get(
 
         return deserialized
     get.metadata = {'url': '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.HanaOnAzure/hanaInstances/{hanaInstanceName}'}
+
+    def restart(
+            self, resource_group_name, hana_instance_name, custom_headers=None, raw=False, **operation_config):
+        """The operation to restart a SAP HANA instance.
+
```
Comment (lagalbra) [inline azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:245]: Add a little bit more description here (get has an extended description of "Gets properties of a SAP HANA instance for the specified subscription, resource group, and instance name.")

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py [hunk 6]
```diff
@@ -237,3 +238,56 @@ def get(
...
+        if response.status_code not in [202]:
+            exp = CloudError(response)
+            exp.request_id = response.headers.get('x-ms-request-id')
+            raise exp
+
+        if raw:
+            client_raw_response = ClientRawResponse(None, response)
+            return client_raw_response
+    restart.metadata = {'url': '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.HanaOnAzure/hanaInstances/{hanaInstanceName}/restart'}
```
Comment (lagalbra) [inline azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:293]: You need a final return line for the function, don't you?

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py [hunk 7]
```diff
@@ -237,3 +238,56 @@ def get(
...
+        request = self._client.post(url, query_parameters, header_parameters)
+        response = self._client.send(request, stream=False, **operation_config)
+
+        if response.status_code not in [202]:
+            exp = CloudError(response)
+            exp.request_id = response.headers.get('x-ms-request-id')
+            raise exp
+
+        if raw:
+            client_raw_response = ClientRawResponse(None, response)
```
Comment (lagalbra) [inline azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:291]: `None` as a parameter makes me nervous; why are you unmarshalling something into nothing?

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py [hunk 8]
```diff
@@ -237,3 +238,56 @@ def get(
...
+            header_parameters.update(custom_headers)
+        if self.config.accept_language is not None:
+            header_parameters['accept-language'] = self._serialize.header("self.config.accept_language", self.config.accept_language, 'str')
+
+        # Construct and send request
+        request = self._client.post(url, query_parameters, header_parameters)
+        response = self._client.send(request, stream=False, **operation_config)
+
+        if response.status_code not in [202]:
+            exp = CloudError(response)
```
Comment (lagalbra) [inline azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:286]: Why does this use something different than `models.ErrorResponseException` like Get does?

**Review by lagalbra — CHANGES_REQUESTED** [review]

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py (hunk 5 above)
Comment (PakDLiu) [reply azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:245]: This is auto generated

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py (hunk 6 above)
Comment (PakDLiu) [reply azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:293]: auto generated

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py (hunk 7 above)
Comment (PakDLiu) [reply azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:291]: auto generated

File: azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py (hunk 8 above)
Comment (PakDLiu) [reply azext_hanaonazure/modules_sdk/operations/hana_instances_operations.py:286]: auto generated

File: HISTORY.rst (hunk 2 above)
Comment (PakDLiu) [reply HISTORY.rst]: done

File: azext_hanaonazure/commands.py (hunk 4 above)
Comment (PakDLiu) [reply azext_hanaonazure/commands.py]: done

File: azext_hanaonazure/azext_metadata.json (hunk 3 above)
Comment (PakDLiu) [reply azext_hanaonazure/azext_metadata.json:4]: that would look weird...

**Review by jeffaco — COMMENTED** [review]
You are going to squash those commits before pushing, right?

Thoughts about a unit test framework?

**Review by lagalbra — APPROVED** [review]

**Comment by PakDLiu** [comment]
About the unit test framework, I think we will need to some day.
For now, the extension only calls the python SDK and has very little actual logic we write.
Also, only the HANA RP Devs are contributing to this right now, I'd image we'd need automated unit test execution when other people start contributing.

For now, I'll add a work item and add it to the backlog.
https://msazure.visualstudio.com/One/_workitems/edit/3022700

**Pull request merged**